src/target/ext/libraries.rs

```diff
@@ -0,0 +1,50 @@
+//! TODO
+
+use crate::target::Target;
+use crate::target::TargetResult;
+
+/// Target Extension - List a target's libraries.
+pub trait Libraries: Target {
+    /// Get library list XML for this target.
+    ///
+    /// See the [GDB Documentation] for a description of the format.
+    ///
+    /// [GDB Documentation]: https://sourceware.org/gdb/current/onlinedocs/gdb.html/Library-List-Format.html
+    ///
+    /// Return the number of bytes written into `buf` (which may be less than
+    /// `length`).
+    ///
+    /// If `offset` is greater than the length of the underlying data, return
+    /// `Ok(0)`.
+    fn get_libraries(
+        &self,
+        offset: u64,
+        length: usize,
+        buf: &mut [u8],
+    ) -> TargetResult<usize, Self>;
+}
+
+define_ext!(LibrariesOps, Libraries);
+
+/// Target Extension - List an SVR4 (System-V/Unix) target's libraries.
+pub trait LibrariesSvr4: Target {
+    /// Get library list XML for this target.
+    ///
+    /// See the [GDB Documentation] for a description of the format.
+    ///
+    /// [GDB Documentation]: https://sourceware.org/gdb/current/onlinedocs/gdb.html/Library-List-Format-for-SVR4-Targets.html
+    ///
+    /// Return the number of bytes written into `buf` (which may be less than
+    /// `length`).
+    ///
+    /// If `offset` is greater than the length of the underlying data, return
+    /// `Ok(0)`.
+    fn get_libraries_svr4(
```

Comment:
Maybe it's not as part of this PR... but having another API force the user return "raw XML" (vs. having `gdbstub` generate it on-the-fly) seems a little unfortunate from a usability POV.

e.g: In this case, I wonder if we could instead have an API that looks something like this instead:

```rust
fn get_libraries_srv4(
    &self, 
    main_lm: Target::Arch::Usize,
    report_lib: &mut dyn FnMut(
        /*name:*/ &str, 
        /*lm:*/ Target::Arch::Usize,
        /*l_addr:*/ Target::Arch::Usize,
        /*l_ld:*/ Target::Arch::Usize, 
        /*lmid:*/ Option<Target::Arch::Usize>
    )
) -> TargetResult<usize, Self>;
```

Though, as I typed this out, I remembered that I basically had the same through process while reviewing #54, and after reviewing that thread... I think the implementation in `gdbstub` would get a bit too hairy (especially when considering how we'd need to handle non-zero `offset` requests...)

Maybe a middle ground could be to keep the "raw XML" API, but also include some helper XML builders that would make it easier for end-users to generate the properly formatted XML they need to return?

e.g:

```rust
struct LibrariesSvr4XmlBuilder<'a, T> { ... }

impl<'a, T> LibrariesSvr4XmlBuilder<'a, T> where T: Target {
    pub fn new_fixed(buf: &'a mut [u8]) -> Self { ... }
    // using a growable vec
    #[cfg(feature = "alloc")]
    pub fn new_dynamic(&mut Vec<u8>) -> Self { ... }
    // returns error if buffer is full
    pub fn add_lib(&mut self, name: &str, lm: T::Arch::Usize, l_addr: T::Arch::Usize, ...) -> TargetResult<(), T> { ... }
}
```

Under the hood, this could re-use the `ManagedVec` type https://github.com/daniel5151/gdbstub/blob/master/src/util/managed_vec.rs

Hmmm...

* * *

...but in any case, I don't think this is something that needs to happen as part of this PR.

I think adding this (and other) XML builders would be a workstream in its own right, and some more thinking is required on what I'd want maintain a cohesive-ish API across all the different kinds of XML that'd be getting generated.

TL;DR: feel free to ignore this comment - I basically just wrote it for future me 😅

Reply:
I think that this is a really good idea yeah, and I was wondering how bad it would be to prototype this. I like your idea of a `LibrariesSvr4XmlBuilder` as a sort of maximally general way of doing this so if raw XML were available it could be given but otherwise it's just as convenient to use the builder. The builder could also have a helper function to take all the raw parameters to `get_libraries_svr4` plus a closure and do it all on-the-fly.